### ISSUE TITLE
Using XLM for BERT-base pertaining

### ISSUE THREAD

**Issue by Shuailong** [facebookresearch/XLM#140]
I used the configurations below to train a BERT-base model from scratch, using the 2015 Wikipedia dump. After 4.5 days of training on a 2-GeForce GTX 1080 GPU card machine, I can get the result shown below. The perplexity on train set increases afterwards, from 2.xx to 7.xxx, and the validation and test perplexity also increases a lot. Is it normal?

```bash
#!/bin/zsh
export NGPU=2
python -m torch.distributed.launch --nproc_per_node=$NGPU train.py \
--exp_name xlm_en_wiki2015 \
--dump_path ./dumped \
--data_path data/processed/XLM_en/30k \
--fp16 true \
--amp 1 \
--lgs 'en' \
--clm_steps '' \
--mlm_steps 'en' \
--emb_dim 768 \
--n_layers 12 \
--n_heads 12 \
--dropout 0.1 \
--attention_dropout 0.1 \
--gelu_activation true \
--batch_size 15 \
--bptt 256 \
--optimizer adam,lr=0.0001 \
--epoch_size 300000 \
--max_epoch 100000 \
--validation_metrics _valid_en_mlm_ppl \
--stopping_criterion _valid_en_mlm_ppl,25 \
--accumulate_gradients 2 \
--word_mask_keep_rand '0.8,0.1,0.1' \
--word_pred '0.15' \
```

> INFO - 07/21/19 14:40:26 - 4 days, 16:54:31 - ============ End of epoch 44 ============
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - epoch -> 44.000000
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - valid_en_mlm_ppl -> 10.860106
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - valid_en_mlm_acc -> 54.600792
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - valid_mlm_ppl -> 10.860106
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - valid_mlm_acc -> 54.600792
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - test_en_mlm_ppl -> 10.671692
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - test_en_mlm_acc -> 54.371764
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - test_mlm_ppl -> 10.671692
> INFO - 07/21/19 14:41:12 - 4 days, 16:55:17 - test_mlm_acc -> 54.371764
>

 ... ...

> 
> INFO - 07/22/19 15:39:20 - 5 days, 17:53:25 - ============ End of epoch 54 ============
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - epoch -> 54.000000
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - valid_en_mlm_ppl -> 1643.229607
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - valid_en_mlm_acc -> 5.363769
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - valid_mlm_ppl -> 1643.229607
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - valid_mlm_acc -> 5.363769
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - test_en_mlm_ppl -> 1693.954061
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - test_en_mlm_acc -> 5.438022
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - test_mlm_ppl -> 1693.954061
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - test_mlm_acc -> 5.438022
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - __log__:{"epoch": 54, "valid_en_mlm_ppl": 1643.229607276697, "valid_en_mlm_acc": 5.363769022305608, "valid_mlm_ppl": 1643.229607276697, "valid_mlm_acc": 5.363769022305608, "test_en_mlm_ppl": 1693.9540612278017, "test_en_mlm_acc": 5.438021768995033, "test_mlm_ppl": 1693.9540612278017, "test_mlm_acc": 5.438021768995033}
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - Not a better validation score (9 / 25).
> INFO - 07/22/19 15:40:07 - 5 days, 17:54:12 - Saving checkpoint to ./dumped/xlm_en_wiki2015/82sdzow6k1/checkpoint.pth ...
> WARNING - 07/22/19 15:40:07 - 5 days, 17:54:12 - Saving model parameters ...
> WARNING - 07/22/19 15:40:07 - 5 days, 17:54:12 - Saving model optimizer ...

> INFO - 07/22/19 15:40:08 - 5 days, 17:54:14 - ============ Starting epoch 55 ... ============
> INFO - 07/22/19 15:40:11 - 5 days, 17:54:17 - 1100005 -    0.78 sent/s -    31.28 words/s - MLM-en:  7.3344 -  - model LR: 1.0000e-04
> INFO - 07/22/19 15:40:13 - 5 days, 17:54:19 - 1100010 -   18.09 sent/s -   727.19 words/s - MLM-en:  7.4152 -  - model LR: 1.0000e-04
> INFO - 07/22/19 15:40:15 - 5 days, 17:54:21 - 1100015 -   17.67 sent/s -   685.47 words/s - MLM-en:  7.4668 -  - model LR: 1.0000e-04

**Comment by glample** [comment]
This is due to some instability issue, this happens when training large models (12 layers is already a large model). There are several ways to make the training more robust:
- use batches as large as possible, or equivalently, as many GPUs as possible
- if you don't have many GPUs, you can increase the value `--accumulate_gradients` to simulate larger batches
- use a better optimizer / learning rate schedule. something like `--optimizer adam_inverse_sqrt,lr=0.0001,warmup_updates=20000,beta1=0.9,beta2=0.999,weight_decay=0.01,eps=0.000001` should do the trick

There are several important parameters in the optimizer:
`beta2=0.999`, `eps=1e-6`, it works better than default value of `1e-8` (it's also more stable). Also, `weight_decay=0.01`, it prevents the model weights from diverging. In particular, without it we found that some parameters become extremely large, especially in the self-attention layers.

Hope this helps!

**Comment by Shuailong** [comment]
> This is due to some instability issue, this happens when training large models (12 layers is already a large model). There are several ways to make the training more robust:
> 
> * use batches as large as possible, or equivalently, as many GPUs as possible
> * if you don't have many GPUs, you can increase the value `--accumulate_gradients` to simulate larger batches
> * use a better optimizer / learning rate schedule. something like `--optimizer adam_inverse_sqrt,lr=0.0001,warmup_updates=20000,beta1=0.9,beta2=0.999,weight_decay=0.01,eps=0.000001` should do the trick
> 
> There are several important parameters in the optimizer:
> `beta2=0.999`, `eps=1e-6`, it works better than default value of `1e-8` (it's also more stable). Also, `weight_decay=0.01`, it prevents the model weights from diverging. In particular, without it we found that some parameters become extremely large, especially in the self-attention layers.
> 
> Hope this helps!

Thank you very much for your suggestions!

1. I try to use batch-size of 15 and accumulate steps of 2, to make an effective batch size of 30, which is near the recommended 32. Can I increase the accumulated steps to 5, 10, etc, any side effect about this way of increasing batch size?

2. Could you please give me any hints on what train/validation perplexity value is good enough? 

Thank you again for your time!

Regards,
Shuailong

**Comment by glample** [comment]
I recommend using multiples of 8 for the batch size, if you have float16 it will be much faster (the code should take care of padding dimensions to multiples of 8 though, but you can directly set 16 instead of 15). Increasing accumulated steps to 8 should be fine, yes.

Regarding the expected perplexity, I'm not sure, this really depends on the dataset. But 60% or 70% word prediction accuracy should be pretty good.

**Comment by Shuailong** [comment]
> I recommend using multiples of 8 for the batch size, if you have float16 it will be much faster (the code should take care of padding dimensions to multiples of 8 though, but you can directly set 16 instead of 15). Increasing accumulated steps to 8 should be fine, yes.
> 
> Regarding the expected perplexity, I'm not sure, this really depends on the dataset. But 60% or 70% word prediction accuracy should be pretty good.

Great! Thanks!
